File: 351002/Yakhnovets/rest-task/src/issue/controller/issue.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  ParseIntPipe,
  Post,
  Put,
} from '@nestjs/common';
import { IssueService } from '../service/issue.service';
import { IssueRequestTo } from '../dto/issue-request.to';
import { IssueResponseTo } from '../dto/issue-response.to';

@Controller('api/v1.0/issues')
export class IssueController {
  constructor(private readonly service: IssueService) {}

  @Get()
  getAll(): Promise<IssueResponseTo[]> {
    return this.service.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseIntPipe) id: number): Promise<IssueResponseTo> {
    return this.service.getById(id);
  }

  @Post()
  create(@Body() dto: IssueRequestTo): Promise<IssueResponseTo> {
    return this.service.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: IssueRequestTo,
  ): Promise<IssueResponseTo> {
    return this.service.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(204)
  delete(@Param('id', ParseIntPipe) id: number): Promise<void> {
    return this.service.delete(id);
  }
}

File: 351004/Student/Lab6/publisher/src/main/java/com/example/task310/controller/IssueController.java
package com.example.task310.controller;

import com.example.task310.dto.*;
import com.example.task310.service.IssueService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping({"/api/v1.0/issues", "/api/v2.0/issues"})
@RequiredArgsConstructor
public class IssueController {
    private final IssueService service;

    @GetMapping
    public List<IssueResponseTo> getAll(Pageable pageable) {
        return service.getAll(pageable);
    }

    @GetMapping("/{id}")
    public IssueResponseTo getById(@PathVariable Long id) {
        return service.getById(id);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    @PreAuthorize("isAnonymous() or @accessGuard.canManageIssueByWriter(#dto.writerId, authentication)")
    public IssueResponseTo create(@Valid @RequestBody IssueRequestTo dto) {
        return service.create(dto);
    }

    @PutMapping("/{id}")
    @PreAuthorize("isAnonymous() or @accessGuard.canManageIssue(#id, authentication)")
    public IssueResponseTo update(@PathVariable Long id, @Valid @RequestBody IssueRequestTo dto) {
        dto.setId(id);
        return service.update(dto);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @PreAuthorize("isAnonymous() or @accessGuard.canManageIssue(#id, authentication)")
    public void delete(@PathVariable Long id) {
        service.delete(id);
    }
}
